src/kernel/sporadic.c

```diff
@@ -210,135 +214,84 @@ void refill_update(sched_context_t *sc, ticks_t new_period, ticks_t new_budget,
-         * rAmount in a refill must be >= MIN_BUDGET */
+    if (!refill_single(sc)) {
+        ticks_t amount = refill_head(sc)->rAmount;
+        ticks_t tail = refill_head(sc)->rTime + amount;
+        return refill_index(sc, refill_next(sc, sc->scRefillHead))->rTime <= tail;
+    } else {
+        return false;
     }
 }
 
 void refill_budget_check(ticks_t usage)
 {
     sched_context_t *sc = NODE_STATE(ksCurSC);
-    /* this function should only be called when the sc is out of budget */
-    ticks_t capacity = refill_capacity(NODE_STATE(ksCurSC), usage);
-    assert(capacity < MIN_BUDGET || refill_full(sc));
-    assert(sc->scPeriod > 0);
+    assert(!isRoundRobin(sc));
     REFILL_SANITY_START(sc);
 
-    if (capacity == 0) {
-        while (refill_head(sc)->rAmount <= usage) {
-            /* exhaust and schedule replenishment */
-            usage -= refill_head(sc)->rAmount;
-            if (refill_single(sc)) {
-                /* update in place */
-                refill_head(sc)->rTime += sc->scPeriod;
-            } else {
-                refill_t old_head = refill_pop_head(sc);
-                old_head.rTime = old_head.rTime + sc->scPeriod;
-                schedule_used(sc, old_head);
-            }
-        }
-
-        /* budget overrun */
-        if (usage > 0) {
-            /* budget reduced when calculating capacity */
-            /* due to overrun delay next replenishment */
-            refill_head(sc)->rTime += usage;
-            /* merge front two replenishments if times overlap */
-            if (!refill_single(sc) &&
-                refill_head(sc)->rTime + refill_head(sc)->rAmount >=
-                refill_index(sc, refill_next(sc, sc->scRefillHead))->rTime) {
-
-                refill_t refill = refill_pop_head(sc);
-                refill_head(sc)->rAmount += refill.rAmount;
-                refill_head(sc)->rTime = refill.rTime;
-            }
-        }
+    /* We assume that the estimates on worst-case kernel overheads are
+     * correct and that overrun cannot occur. */
+#ifdef CONFIG_DEBUG_BUILD
+    if (usage > refill_head(sc)->rAmount) {
+        ticks_t overrun = usage - refill_head(sc)->rAmount;
+        userError(
+            "SC (%p) overran budget by %lluus (%llu ticks)",
+            sc,
+            ticksToUs(overrun),
+            overrun
+        );
     }
-
-    capacity = refill_capacity(sc, usage);
-    if (capacity > 0 && refill_ready(sc)) {
-        refill_split_check(usage);
+#endif
+    usage = MIN(usage, refill_head(sc)->rAmount);
+
+    /* Charge the usage to the head refill */
+    if (usage > 0) {
+        refill_t used = (refill_t) {
+            .rAmount = usage,
+            .rTime = refill_head(sc)->rTime + sc->scPeriod,
+        };
+
+        refill_head(sc)->rAmount -= usage;
+        /* We need to keep the head refill no more than a period before
+         * the start of the tail refill. */
+        refill_head(sc)->rTime += usage;
+        schedule_used(sc, used);
```

Comment:
From a scheduling theory PoV, this should be necessary only when the thread either blocks (i.e., voluntary relinquishes the processor) or if it runs out of budget. This shouldn't be done while the thread remains runnable and has sufficient budget to keep executing.

I tried figuring out the call-chain via `chargeBudget()` and `checkBudget()`, but couldn't determine whether `refill_budget_check()` is called only on blocking threads / threads with expired budget (the names suggests otherwise). Could you please explain whether it's called every time the scheduler considers the currently running TCB?

Reply:
`refill_budget_check` is called whenever we switch from one thread to another; this could be because of budget exhaustion or because the thread has blocked, but it could also be the result of preemption. I suspect the reason why this is also called in the case of preemption is that we don't track the amount of time _consumed_ or _avaialbe_ for an SC other than in the replenishment queue itself. The head refill indicated time available (if its start time is in the past) and time used is tracked by deducting that from the head release and scheduling it as a new refill.

As for how this is achieved if `checkBudget` returns `true` on entry, the SC has sufficient budget for a kernel invocation and has not already been charged for any cumulative time spent executing. In `schedule` transitively calls `commitTime` (indicating that we are not returning to a thread using the same SC as the one used on entry into the kernel) we charge the current SC if there is any accumulated consumed budget outstanding.

Reply:
Just to make sure I understand: from a real-time PoV, a preemption should *not* trigger a separate refill to be scheduled, but this code path actually does so. However, the [above-discussed code](https://github.com/seL4/seL4/pull/256#discussion_r527358278) causes these superfluous replenishments to be merged and hence elided, so essentially it doesn't have an effect on the actual replenishment points and amounts. Is that a correct understanding of how this works out?

Reply:
Yes, in the case of preemtion, the above code ensures that rather than a new refill being created the time is instead added to the refill created before the preemption. The effect is that time from the head refill is moved to the new refill one period into the future incrementally with each preemption.